packages/shared/utils/environment.ts

```diff
@@ -25,103 +35,198 @@ export function isReplayBrowser() {
+
+        updateUrlToReflectCurrentState();
+      }
+    },
+  });
 
-export function getUrlString(params: Record<string, string | null | undefined>) {
-  const url = new URL(window.location.toString());
+  startAppListening({
+    actionCreator: paused,
+    effect: async ({ payload: { executionPoint, time } }) => {
+      if (mutableParams.point !== executionPoint || mutableParams.time !== time) {
+        mutableParams.point = executionPoint;
+        mutableParams.time = time;
 
-  Object.entries(params).forEach(([key, value]) => {
-    if (value !== undefined && value !== null) {
-      url.searchParams.set(key, value);
-    }
+        // Changing the current time should also reset/unset the selected comment
+        mutableParams.commentId = null;
+
+        updateUrlToReflectCurrentState();
+      }
+    },
   });
 
-  return url.toString();
-}
+  startAppListening({
+    actionCreator: setTimelineState,
+    effect: async ({ payload: { currentTime } }) => {
+      if (currentTime != null && mutableParams.time !== currentTime) {
+        mutableParams.point = null;
+        mutableParams.time = currentTime;
 
-export function updateUrlWithParams(params: Record<string, string | null | undefined>) {
-  const urlString = getUrlString(params);
+        // Changing the current time should also reset/unset the selected comment
+        mutableParams.commentId = null;
 
-  window.history.replaceState({}, "", urlString);
-}
+        updateUrlToReflectCurrentState();
+      }
+    },
+  });
 
-export function getLoginReferrerParam() {
-  const url = getURL();
-  const referrerParam = url.searchParams.get("login-referrer");
-  return referrerParam === "first-browser-open" ? referrerParam : "default";
+  startAppListening({
+    actionCreator: setFocusWindow,
+    effect: async () => {
+      // This action signals a change in the focus window,
+      // but the payload only contains times (numbers).
+      // ReplayClient contains the higher resolution execution points.
+      // The timeline actions awaits ReplayClient before finishing.
+      const focusWindow = replayClient.getCurrentFocusWindow();
+
+      if (mutableParams.focusWindow != focusWindow) {
+        mutableParams.focusWindow = focusWindow;
+
+        updateUrlToReflectCurrentState();
+      }
+    },
+  });
+
+  // TODO Use startAppListening() for these actions (once they're RTK based)
+  store.subscribe(() => {
+    const state = store.getState();
+
+    const primaryPanel = getSelectedPrimaryPanel(state);
+    const secondaryPanel = getSelectedPanel(state);
+    const viewMode = getViewMode(state);
+
+    if (
+      mutableParams.primaryPanel != primaryPanel ||
+      mutableParams.secondaryPanel != secondaryPanel ||
+      mutableParams.viewMode != viewMode
+    ) {
+      mutableParams.primaryPanel = primaryPanel;
+      mutableParams.secondaryPanel = secondaryPanel;
+      mutableParams.viewMode = viewMode;
+
+      updateUrlToReflectCurrentState();
+    }
+  });
 }
```

Comment:
Clarify the question?

If you're asking "how can I use the listener middleware to handle actions that are plain strings and were not defined in an RTK `createSlice` call", the answer is you can just pass the type string directly to `startAppListening` as the `type` option:

```ts
startAppListening({
  type: "set_selected_panel",
  effect: (action, listenerApi) => {
    const state = listenerApi.getState();
    // more logic
  }
})
```

On a related note, it's hypothetically technically more efficient to add one listener that checks for multiple action types (so there's only one listener entry that gets checked on each dispatched action) rather than several separate listeners, but it's a micro-optimization and probably not worth worrying about atm.

Reply:
Ah, I didn't know about the `type` param. Thanks!

> On a related note, it's hypothetically technically more efficient to add one listener that checks for multiple action types

To be fair, that's what I would have _preferred_ to do (and considered doing) because of the number of things I am checking here, but I didn't know if that would have performance gotchas. I think I might replace these individual listeners with one debounced subscribe check then.